rclrs/src/node/client.rs

```diff
@@ -0,0 +1,257 @@
+use crate::node::client::oneshot::Canceled;
+use futures::channel::oneshot;
+use std::borrow::Borrow;
+use std::boxed::Box;
+use std::collections::HashMap;
+use std::ffi::CString;
+use std::sync::atomic::{AtomicI64, Ordering};
+use std::sync::Arc;
+
+use crate::error::{RclReturnCode, ToResult};
+use crate::MessageCow;
+use crate::Node;
+use crate::{rcl_bindings::*, RclrsError};
+
+use parking_lot::{Mutex, MutexGuard};
+use rosidl_runtime_rs::Message;
+
+// SAFETY: The functions accessing this type, including drop(), shouldn't care about the thread
+// they are running in. Therefore, this type can be safely sent to another thread.
+unsafe impl Send for rcl_client_t {}
+
+/// Internal struct used by clients.
+pub struct ClientHandle {
+    rcl_client_mtx: Mutex<rcl_client_t>,
+    rcl_node_mtx: Arc<Mutex<rcl_node_t>>,
+}
+
+impl ClientHandle {
+    pub(crate) fn lock(&self) -> MutexGuard<rcl_client_t> {
+        self.rcl_client_mtx.lock()
+    }
+}
+
+impl Drop for ClientHandle {
+    fn drop(&mut self) {
+        let handle = self.rcl_client_mtx.get_mut();
+        let rcl_node_mtx = &mut *self.rcl_node_mtx.lock();
+        unsafe {
+            rcl_client_fini(handle as *mut _, rcl_node_mtx as *mut _);
+        }
+    }
+}
+
+impl From<Canceled> for RclrsError {
+    fn from(_: Canceled) -> Self {
+        RclrsError::RclError {
+            code: RclReturnCode::Error,
+            msg: None,
+        }
+    }
+}
+
+/// Trait to be implemented by concrete Client structs
+/// See [`Client<T>`] for an example
+pub trait ClientBase: Send + Sync {
+    /// Internal function to get a reference to the `rcl` handle.
+    fn handle(&self) -> &ClientHandle;
+    /// Tries to take a new response and run the callback or future with it.
+    fn execute(&self) -> Result<(), RclrsError>;
+}
+
+type RequestValue<Response> = Box<dyn FnOnce(&Response) + 'static + Send>;
+
+/// Main class responsible for publishing data to ROS topics
+pub struct Client<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    pub(crate) handle: Arc<ClientHandle>,
+    requests: Mutex<HashMap<i64, RequestValue<T::Response>>>,
```

Comment:
Might be handy to have a comment or better yet type alias to explain what the `i64` means.

Reply:
Done in 2786288